### ISSUE TITLE
Add behavioral detector via Trace-JEPA: LLMTrace ↔ AutoJEPA bridge

### ISSUE THREAD

**Issue by epappas** [techlab-innov/llmtrace#267]
# Add behavioral detector via Trace-JEPA: LLMTrace ↔ AutoJEPA bridge

## Summary

Add a fifth detector to LLMTrace's existing content-based ensemble (regex,
DeBERTa, InjecGuard, PIGuard) that votes from **action-sequence shape** rather
than prompt content. This is a self-supervised behavioral encoder trained
under [Trace-JEPA](https://github.com/epappas/autojepa/blob/master/examples/trace-jepa/research.md)
in the sibling [AutoJEPA](https://github.com/epappas/autojepa) project.

Crossed the Phase-3 falsifier on synthesized data at
`probe_auroc = 0.7516` at 5% FPR (v0.2.0,
[evidence](https://github.com/epappas/autojepa/tree/master/artifacts/trace-jepa)).
The single missing experiment is **does the signal transfer to real
LLMTrace traces?** — this issue scopes the work to answer that question and
ship the detector if the answer is yes.

## Why this matters for LLMTrace

LLMTrace's current detector ensemble is content-based: every detector reads
the agent's **prompts / inputs / outputs** and pattern-matches against attack
shapes. The ensemble hits **87.6% accuracy / 95.5% precision / 86.9% F1 /
79.7% recall** on the published 153-sample adversarial corpus — that's
strong but it leaves one structural blind spot: **adversarial paraphrasing
of the malicious prompt degrades every content detector simultaneously**
(they share the same input modality).

A behavioral detector that votes from the action-sequence shape (e.g.
`write_file → shell_exec → fetch_url` is anomalous regardless of what string
the agent thinks it's executing) is independent of paraphrasing. If we can
show even a moderately-accurate behavioral signal is statistically
independent of the content ensemble's errors, adding it strictly improves
ensemble PR at the same FPR.

This is the long-term moat against adapting attackers: content-detector arms
race is asymmetric (attackers iterate strings cheaply), behavioral-detector
arms race requires the attacker to change the *agent's actions*, which is
much harder to disguise.

## What's already done (in sibling repo)

- **Detector trained**: 116 MB checkpoint at
  [`artifacts/trace-jepa/runs/run-0003/model/best.pt`](https://github.com/epappas/autojepa/tree/master/artifacts/trace-jepa/runs/run-0003/model)
  in autojepa (Git LFS). Inference takes a Session JSON and returns a scalar
  attack probability.
- **Schema alignment confirmed**: AutoJEPA's per-event schema
  (`action_name`, `action_type`, `args`, `return_code`, `timestamp`,
  `actor_id`, `parent_link`) was designed to match what LLMTrace already
  collects. No tokenizer redesign needed.
- **Reproducible corpus**: 355 MB of synthesized training shards committed
  via Git LFS at
  [`examples/trace-jepa/data/`](https://github.com/epappas/autojepa/tree/master/examples/trace-jepa/data).
  Anyone with `git lfs pull` has the exact training distribution.
- **Full research writeup**:
  [examples/trace-jepa/research.md](https://github.com/epappas/autojepa/blob/master/examples/trace-jepa/research.md)
  (with charts, methodology, attribution caveats, and 7-category caveats list).

## Scope of this issue

Five concrete deliverables, in the order they unblock each other:

### 1. LLMTrace trace export (~2 days)

**Goal**: dump real production traces in the schema AutoJEPA already consumes.

Tasks:
- [ ] Add an export endpoint or batch CLI to `llmtrace-core` that emits per-session JSON in the form:
  ```json
  {
    "session_id": <int>,
    "events": [
      {
        "action_name": "search_web",
        "action_type": "tool_call",
        "args": "query=...",
        "return_code": 0,
        "timestamp": 0.42,
        "actor_id": 2,
        "parent_link": 1
      }
    ],
    "is_attack": false
  }
  ```
- [ ] Each event must map from LLMTrace's internal event types:
  - LLM API calls → `action_type=llm_call`
  - Tool calls → `action_type=tool_call`
  - Tool returns → `action_type=return`
  - Plan steps / observations if instrumented → `action_type=plan_step / observation`
- [ ] `is_attack` label drawn from LLMTrace's current detector ensemble's
  majority vote (so we can train the new detector against the labels the
  current ensemble produces; mismatches are the new detector's potential
  value-add).
- [ ] Output format: gzipped JSONL or WebDataset `.tar` shards (same format
  AutoJEPA's `prepare.py` already reads).
- [ ] CLI flag for time range (`--since 2026-04-01 --until 2026-05-01`) so
  we can replay any month of past traffic.

**Acceptance**: a 100-session export round-trips through
`autojepa/examples/trace-jepa/prepare.py` (with the `--from-llmtrace` mode
below) without schema errors.

### 2. AutoJEPA real-trace ingestion mode (~1 day, lives in autojepa repo)

**Goal**: let `prepare.py` read LLMTrace exports instead of synthesizing.

Tasks (in [autojepa](https://github.com/epappas/autojepa) repo, separate PR):
- [ ] Add `--from-llmtrace <path>` flag to
  `examples/trace-jepa/prepare.py`. When set, skip synthesis and read shards
  from the LLMTrace export directory.
- [ ] Preserve the train/probe/canary split structure (manifest.json should
  look identical regardless of source).
- [ ] Document the schema mapping in `examples/trace-jepa/README.md`.

**Acceptance**: `python prepare.py --from-llmtrace path/to/export/` produces
the same on-disk layout as `python prepare.py --n-sessions 100000`.

### 3. Re-run the trace-jepa campaign on real traces (~6 GPU-hours, ~$15 on Basilica)

**Goal**: measure whether the synth→real transfer holds.

Tasks:
- [ ] Export 100k real sessions from LLMTrace via step 1.
- [ ] Run AutoJEPA's hybrid campaign against the real corpus (same 30-iter
  budget, same hybrid policy, same model architecture).
- [ ] Record headline metrics: best `probe_auroc`, kept-iter rate,
  reproducibility across seeds.
- [ ] **Critical decision point**: does `probe_auroc > 0.70` hold on real
  traces?
  - If **yes**: proceed to step 4 (ensemble integration).
  - If **no**: write a short post-mortem identifying what shape of real
    behavior the synthesizer was missing; revise `prepare.py`; iterate.

**Acceptance**: a published 30-iter campaign log with a clear yes/no on the
KPI threshold and the new `best.pt` checkpoint if positive.

### 4. Ensemble integration (~2 days, lives partly in llmtrace-security, partly in llmtrace-core)

**Goal**: wire trace-jepa as the fifth voter in LLMTrace's existing ensemble.

Tasks:
- [ ] Add a `trace_jepa` detector to `crates/llmtrace-security/` that loads
  the trained encoder + linear probe and returns `f64` attack probability
  per session. PyTorch model loaded via `tch` or `candle`; alternatively
  via a sidecar Python subprocess if Rust torch bindings are too heavy.
- [ ] Detector interface should mirror the existing four content detectors
  (regex, DeBERTa, InjecGuard, PIGuard) — same `detect(session) -> Score`
  contract so it slots into the majority-voting layer without changes.
- [ ] Wire it into `LLMTraceProxy`'s detection pipeline under a feature
  flag (`config.security.detectors.trace_jepa.enabled: bool`, default off
  until calibrated).
- [ ] Per-session inference target: <50ms p99 on CPU, <10ms on GPU. If
  these aren't met on real production batch shapes, document and consider
  batching.

**Acceptance**: a new detector entry in LLMTrace's existing benchmark
output:
```
| Detector       | Accuracy | Precision | F1    | Recall |
|---|---|---|---|---|
| Ensemble (4)   |   87.6%  |   95.5%   | 86.9% | 79.7%  |
| trace_jepa     |    ?     |     ?     |   ?   |    ?   |
| Ensemble (5)   |    ?     |     ?     |   ?   |    ?   |
```

### 5. Measure ensemble independence (~half day analysis, no new code)

**Goal**: validate the load-bearing assumption that trace-jepa's errors
are statistically independent of the content detectors' errors.

Tasks:
- [ ] Run both the existing 4-detector ensemble and trace-jepa on the same
  153-sample (or larger) adversarial corpus.
- [ ] Compute the **error-correlation matrix** — for each pair of
  detectors, what fraction of attacks they both miss vs miss independently.
- [ ] Decision criteria:
  - If max pairwise error correlation between trace_jepa and any existing
    detector is **< 0.5**: ship the ensemble update (real independent
    signal).
  - If correlation is **0.5–0.8**: ship behind a feature flag, document
    that it provides some redundancy and marginal independence.
  - If correlation is **> 0.8**: do not ship. The new detector is
    structurally similar to one of the existing four; the ensemble gain
    is minimal and not worth the deployment cost.

**Acceptance**: the error-correlation matrix in a PR description, with a
clear ship/no-ship decision.

## Estimated total effort

- **Engineering**: ~1 week of focused work (steps 1–2: 3 days; step 4:
  2 days; steps 3 + 5: mostly automated).
- **GPU**: ~$25 of Basilica A100 time (step 3 main campaign + a paired
  attribution ablation in the autojepa repo).
- **Calendar**: 2–3 weeks elapsed including review cycles.

## Outcomes (either is useful)

**If the signal transfers and the ensemble gains**: LLMTrace ships its first
behavioral detector. Marketing claim becomes "five-detector ensemble with
content + behavioral defense in depth." Strong story for prospects who
worry about adversarial paraphrasing.

**If the signal doesn't transfer or correlates too strongly with existing
detectors**: we've learned the most important thing — what about the real
trace distribution breaks the synthesized assumption, OR whether
behavioral and content signals are actually independent in production
data. Either answer is a defensible research result and a clear next-step
direction.

## What's explicitly NOT in scope

- **Replacing the existing content detectors.** trace-jepa is additive.
- **Live retraining loop** (continuous adaptation as production traffic
  evolves). That's a follow-up issue after we've validated the static
  detector.
- **Multi-deployment support** (one model per LLMTrace customer with their
  own normal-behavior baseline). Follow-up after this issue confirms the
  basic deployment story.
- **Benchmark against published baselines** (LogLLaMA, GraphIDS, SAFE, MTS-
  JEPA). That's a research paper, not a product issue; tracked separately in
  [autojepa#TBD](https://github.com/epappas/autojepa/issues).

## Cross-repo coordination

This issue lives in `techlab-innov/llmtrace` because the integration is the
load-bearing piece. The companion work in
[`epappas/autojepa`](https://github.com/epappas/autojepa) is steps 2 and 3
(prepare.py ingestion mode + the real-trace campaign) and will be filed as
a separate issue there once this one is accepted. Both should be tracked
under the same milestone.

## Risk register

| Risk | Likelihood | Impact | Mitigation |
|---|---|---|---|
| Real-trace AUROC drops below 0.70 | Medium | High (kills the deployment story) | Step 3 post-mortem identifies the synth gap; iterate on `prepare.py` |
| Error correlation > 0.8 | Low | Medium (kills the ensemble argument) | Step 5 is explicitly an early-exit gate; we don't ship a redundant detector |
| Rust-side torch integration too heavy | Medium | Low (can fall back to Python sidecar) | Step 4 has a `tch / candle / Python sidecar` decision tree; prototype both and pick winner |
| LLMTrace schema drift from autojepa schema | Low | Medium (would require synth-corpus rewrite) | Step 1 acceptance test catches this on day 1 |

## References

- AutoJEPA: https://github.com/epappas/autojepa (v0.2.0)
- Trace-JEPA research writeup: https://github.com/epappas/autojepa/blob/master/examples/trace-jepa/research.md
- Trained checkpoint: https://github.com/epappas/autojepa/tree/master/artifacts/trace-jepa/runs/run-0003/model
- Trace schema source-of-truth: https://github.com/epappas/autojepa/blob/master/examples/trace-jepa/prepare.py
- ADR-029 (the on-improve checkpoint mechanism that makes the model deployable): https://github.com/epappas/autojepa/blob/master/docs/adr/029-train-py-best-state-checkpoint-on-improve.md

## Labels (suggested)

`enhancement` · `security` · `cross-repo` · `research-driven` · `needs-design-review`
